samples/multiplatform/kmp-lib-sample/src/iosMain/kotlin/com/apollographql/apollo/kmpsample/data/ApolloNetworkClient.kt

```diff
@@ -0,0 +1,145 @@
+      assert(NSThread.isMainThread())
+      val continuationRef = StableRef.create(continuation).asCPointer()
+      try {
+        val delegate = { httpData: NSData?, httpResponse: NSURLResponse?, error: NSError? ->
+          initRuntimeIfNeeded()
+          val response = parse(
+              data = httpData,
+              httpResponse = httpResponse as NSHTTPURLResponse,
+              error = error
+          )
+          response.dispatchOnMain(continuationRef)
+        }
+        session.dataTaskWithRequest(request.freeze(), delegate.freeze()).resume()
+      } catch (e: Exception) {
+        continuationRef.asStableRef<CancellableContinuation<Response<T>>>().dispose()
+        continuation.resumeWithException(e)
+      }
+    }
+  }
+
+  private fun <D : Operation.Data> Operation<D, *, *>.prepareRequest(
+      url: String,
+      headers: Map<String, String> = emptyMap()
+  ): NSMutableURLRequest {
+    return NSMutableURLRequest.requestWithURL(NSURL(string = url)).apply {
+      setHTTPMethod("POST")
+      setCachePolicy(NSURLRequestReloadIgnoringCacheData)
+      headers.forEach { (key, value) -> setValue(value, forHTTPHeaderField = key) }
+      setValue(operationId(), forHTTPHeaderField = "X-APOLLO-OPERATION-ID")
+      setValue(name().name(), forHTTPHeaderField = "X-APOLLO-OPERATION-NAME")
+      setHTTPBody(composeRequestBody().toByteArray().toNSData())
+    }
+  }
+
+  private fun <T : Operation.Data> Operation<*, T, *>.parse(
+      data: NSData?,
+      httpResponse: NSHTTPURLResponse,
+      error: NSError?
+  ): Result<T> {
+    if (error != null) {
+      return Result.Failure(IOException(error.localizedDescription))
+    }
+
+    return try {
+      val statusCode = httpResponse.statusCode.toInt()
+      if (statusCode in 200..299) {
+        val payload = data!!.toByteArray()
+        val response = parse(Buffer().write(payload))
+        Result.Success(response)
+      } else {
+        Result.Failure(IOException("Network request failed, HTTP status code `$statusCode`"))
+      }
+    } catch (e: Exception) {
+      Result.Failure(e)
+    }
+  }
+
+  private sealed class Result<T : Operation.Data> {
+    class Success<T : Operation.Data>(val value: Response<T>) : Result<T>()
+    class Failure<T : Operation.Data>(val error: Exception) : Result<T>()
+  }
+
+  private fun ByteArray.toNSData(): NSData = NSMutableData().apply {
+    if (isEmpty()) return@apply
+    this@toNSData.usePinned {
+      appendBytes(it.addressOf(0), size.convert())
+    }
+  }
+
+  private fun NSData.toByteArray(): ByteArray {
+    val data: CPointer<ByteVar> = bytes!!.reinterpret()
+    return ByteArray(length.toInt()) { index -> data[index] }
+  }
+
+  @Suppress("NAME_SHADOWING")
+  private fun <T : Operation.Data> Result<T>.dispatchOnMain(continuationRef: COpaquePointer) {
+    dispatch_async_f(
+        queue = dispatch_get_main_queue(),
+        context = DetachedObjectGraph { (continuationRef to this).freeze() }.asCPointer(),
```

Comment:
@sav007 Is there any specific reason to use `DetachedObjectGraph` here compared to just `StableRef` ? I was under the impression that the  advantage of `DetachedObjectGraph` was that it could work with unfrozen references by making sure no one else accesses them but since everything is frozen here, a `StableRef` would work just as well ?

Reply:
I think you are right here, have to double-check but looks like no need for detached object wrapper. Will check

Reply:
```
  @Suppress("NAME_SHADOWING")
  private fun <T : Operation.Data> Result<T>.dispatchOnMain(continuationRef: COpaquePointer) {
    val continuationWithResultRef = StableRef.create((continuationRef to this).freeze())
    dispatch_async_f(
        queue = dispatch_get_main_queue(),
        context = continuationWithResultRef.asCPointer(),
        work = staticCFunction { ref ->
          val continuationWithResultRef = ref!!.asStableRef<Pair<COpaquePointer, Result<T>>>()
          val (continuationPointer, result) = continuationWithResultRef.get()
          continuationWithResultRef.dispose()

          val continuationRef = continuationPointer.asStableRef<CancellableContinuation<Response<T>>>()
          val continuation = continuationRef.get()
          continuationRef.dispose()

          when (result) {
            is Result.Success -> continuation.resume(result.value)
            is Result.Failure -> continuation.resumeWithException(result.error)
          }
        }
    )
  }
```

This works perfectly. Will open PR.